rotary/repair.py

```diff
@@ -748,23 +783,31 @@ def run_end_repair(long_read_filepath, assembly_fasta_filepath, assembly_info_fi
     #   but I imagine it is closer to the behaviour that the user would expect.
     linear_contig_names = parse_assembly_info_file(assembly_info_filepath, assembly_info_type, return_type='linear')
 
-    if keep_failed_contigs is True:
-        linear_contig_names.append(failed_contig_names)
-
     with open(end_repaired_contigs_filepath, 'a') as append_handle:
 
         for record in subset_sequences(assembly_fasta_filepath, linear_contig_names):
             SeqIO.write(record, append_handle, 'fasta')
 
+    # Write info file of how contigs were repaired
+    repaired_contig_names = list(set(circular_contig_names).difference(set(failed_contig_names)))
+    repair_info = pd.concat([pd.DataFrame({'contig': repaired_contig_names, 'circular': 'Y', 'repaired': 'Y'}),
+                             pd.DataFrame({'contig': failed_contig_names, 'circular': 'Y', 'repaired': 'N'}),
+                             pd.DataFrame({'contig': linear_contig_names, 'circular': 'N', 'repaired': 'N'})], axis=0)
+    repair_info.to_csv(end_repair_status_filepath, sep='\t', index=False)
+
     # Clean up temp files
     os.remove(bam_filepath)
     os.remove(f'{bam_filepath}.bai')
     shutil.rmtree(linking_outdir_base)
 
-    logger.info(f'End repair finished. Output contigs saved at {end_repaired_contigs_filepath}.')
-
-    if len(failed_contig_names) != 0:
-        logger.warning(f'{len(failed_contig_names)} contigs could not be circularized - see above for details')
+    logger.info('End repair finished!')
+    logger.info('#### Final stats: ####')
+    logger.info(f'Circular, repaired: {len(repaired_contig_names)} contig(s)')
+    logger.info(f'Circular, not repairable: {len(failed_contig_names)} contig(s)')
+    logger.info(f'Linear: {len(linear_contig_names)} contig(s)')
+    logger.info('######################')
+    logger.info(f'Output contigs are saved at {end_repaired_contigs_filepath}. '
+                f'A summary of repair work is saved at {end_repair_status_filepath}.')
 
 
 def main():
```

Comment:
@jmtsuji I moved move ```main()``` to the top of the file  justbelow the imports. That way, when the user opens the file, they don't have to scroll around searching for ```main()```. I would then order roughly where the functions occur in the file in the order called by ```main()```. Having them in that order makes it easier to look up if you read through the ```main()``` code. The exception to this pattern is that I would leave the argsparse code at the bottom.

Reply:
The order in which the functions are defined doesn't matter in Python, as they are all loaded at runtime.

Reply:
Thanks for the advice -- main() is now at the top of the file